packages/@aws-cdk/aws-batch/lib/compute-environment.ts

```diff
@@ -0,0 +1,484 @@
+  constructor(scope: Construct, id: string, props: ComputeEnvironmentProps = { enabled: true, managed: true }) {
+    super(scope, id, {
+      physicalName: props.computeEnvironmentName,
+    });
+
+    this.validateProps(props);
+
+    const spotFleetRole = this.getSpotFleetRole(props);
+    let computeResources: CfnComputeEnvironment.ComputeResourcesProperty | undefined;
+
+    // Only allow compute resources to be set when using UNMANAGED type
+    if (props.computeResources && !this.isManagedByAWS(props)) {
+      computeResources = {
+        allocationStrategy: props.allocationStrategy || AllocationStrategy.BEST_FIT,
+        bidPercentage: props.computeResources.bidPercentage,
+        desiredvCpus: props.computeResources.desiredvCpus,
+        ec2KeyPair: props.computeResources.ec2KeyPair,
+        imageId: props.computeResources.image && props.computeResources.image.getImage(this).imageId,
+        instanceRole: props.computeResources.instanceRole
+        ? props.computeResources.instanceRole.roleArn
+        : new iam.LazyRole(this, 'Resource-Instance-Role', {
+          assumedBy: new iam.ServicePrincipal('batch.amazonaws.com'),
+        }).roleArn,
+        instanceTypes: this.buildInstanceTypes(props.computeResources.instanceTypes),
+        launchTemplate: props.computeResources.launchTemplate,
+        maxvCpus: props.computeResources.maxvCpus || 256,
+        minvCpus: props.computeResources.minvCpus || 0,
+        placementGroup: props.computeResources.placementGroup && props.computeResources.placementGroup.ref,
+        securityGroupIds: this.buildSecurityGroupIds(props.computeResources.vpc, props.computeResources.securityGroups),
+        spotIamFleetRole: spotFleetRole ? spotFleetRole.roleArn : undefined,
+        subnets: props.computeResources.vpc.selectSubnets(props.computeResources.vpcSubnets).subnetIds,
+        tags: props.computeResources.computeResourcesTags,
+        type: props.computeResources.type || ComputeResourceType.ON_DEMAND,
+      };
+    }
+
+    const computeEnvironment = new CfnComputeEnvironment(this, 'Resource', {
+      computeEnvironmentName: this.physicalName,
+      computeResources,
+      serviceRole: props.serviceRole
+      ? props.serviceRole.roleArn
+      : new iam.LazyRole(this, 'Resource-Service-Instance-Role', {
+        managedPolicies: [
+          iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSBatchServiceRole'),
+        ],
+        assumedBy: new iam.ServicePrincipal('batch.amazonaws.com'),
+      }).roleArn,
+      state: this.isEnabled(props) ? 'ENABLED' : 'DISABLED',
+      type: this.isManagedByAWS(props) ? 'UNMANAGED' : 'MANAGED',
+    });
+
+    if (props.computeResources && props.computeResources.vpc) {
+      this.node.addDependency(props.computeResources.vpc);
+    }
+
+    this.computeEnvironmentArn = this.getResourceArnAttribute(computeEnvironment.ref, {
+      service: 'batch',
+      resource: 'compute-environment',
+      resourceName: this.physicalName,
```

Comment:
Hmm. Above, in the `fromComputeEnvironmentArn`, you say that the name is taken from the `resource` part of your ARN. But it's `resourceName` here. Pretty sure those should be the same :).

Reply:
Im confused on what you need me to change herein.

Reply:
```ts
  public static fromComputeEnvironmentArn(scope: Construct, id: string, computeEnvironmentArn: string): IComputeEnvironment {
    const stack = Stack.of(scope);
    const computeEnvironmentName = stack.parseArn(computeEnvironmentArn).resource;
```

I meant that if ARN is `arn:aws:batch:us-west-2:123456789012:compute-environment/MyComputeEnvName`, `computeEnvironmentName` will be `compute-environment`, not `MyComputeEnvName`.

This behavior should also have an accompanying unit test.

Reply:
I see, so then is there an easy way to make sure that we always get the right name without using `this.physicalName` ?

Reply:
Unfortunately, no :( You pretty much have to rely on unit tests to check this behavior.

Reply:
I see. Well, I notice I am using `.resource` where I should be using `.resourceName` in the `fromXXX` function. I will get that changed. Is that what you eluding to?

Also, I’ll see if we have any unit tests on this...

Reply:
> I see. Well, I notice I am using .resource where I should be using .resourceName in the fromXXX function. I will get that changed. Is that what you eluding to?

Yes.

Reply:
Is there a way I should handle the case when the name is empty? Should we just toss an Error?